### DISCUSSION TITLE
Fonts listed in annotation window dont match fonts in type.xml

### DISCUSSION THREAD

**Discussion by KFH222** [ImageMagick/ImageMagick#6927] [Help]
### ImageMagick version

7.1.1.1

### Operating system, version and so on

Ubuntu linux

### Description

When I try to annotate an image the
font selection has no relationship to
the fonts in the files listed in type.xml

What is see is in the attached screenshot

the fonts listed include sony-fixed, schumacher, misc-fixed and
bitstream - none work as all tiny

I am running

Description:	Ubuntu 22.04.3 LTS
Release:	22.04
Codename:	jammy

I cannot see why the type.xml file is not being read

my directories  ~.config/ImageMagick/
and
/usr/local/etc/ImageMagick-7/  have 

the type files

type-apple.xml type-freefont.xml type-me.xml type-ubuntu.xml
type-windows.xml 
type-dejavu.xml type-ghostscript.xml type-opensans.xml
type-urw-base35.xml type.xml

of which type.xml only uses the 3 listed

type.xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE typemap [
  <!ELEMENT typemap (type)+>
  <!ATTLIST typemap xmlns CDATA #FIXED ''>
  <!ELEMENT type EMPTY>
  <!ATTLIST type xmlns CDATA #FIXED '' encoding NMTOKEN #IMPLIED
    family CDATA #REQUIRED format NMTOKEN #REQUIRED foundry NMTOKEN #REQUIRED
    fullname CDATA #REQUIRED glyphs CDATA #REQUIRED metrics CDATA #REQUIRED
    name NMTOKEN #REQUIRED stretch NMTOKEN #REQUIRED style NMTOKEN #REQUIRED
    version CDATA #IMPLIED weight CDATA #REQUIRED>
]>

<typemap>
   <include file="type-ghostscript.xml" /> 
   <include file="type-dejavu.xml" />
   <include file="type-urw-base35.xml" />
</typemap>

example type-ghostscript.xml has 

<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE typemap [
  <!ELEMENT typemap (type)+>
  <!ATTLIST typemap xmlns CDATA #FIXED ''>
  <!ELEMENT type EMPTY>
  <!ATTLIST type xmlns CDATA #FIXED '' encoding NMTOKEN #IMPLIED
    family CDATA #REQUIRED format NMTOKEN #REQUIRED foundry NMTOKEN #REQUIRED
    fullname CDATA #REQUIRED glyphs CDATA #REQUIRED metrics CDATA #REQUIRED
    name NMTOKEN #REQUIRED stretch NMTOKEN #REQUIRED style NMTOKEN #REQUIRED
    version CDATA #IMPLIED weight CDATA #REQUIRED>
]>

<typemap>
  <type name="AvantGarde-Book" fullname="AvantGarde Book" family="AvantGarde" foundry="URW" weight="400" style="normal" stretch="normal" format="type1" metrics="/usr/share/fonts/type1/gsfonts/a010013l.afm
" glyphs="/usr/share/fonts/type1/gsfonts/a010013l.pfb"/>
  <type name="AvantGarde-BookOblique" fullname="AvantGarde Book Oblique" family="AvantGarde" foundry="URW" weight="400" style="oblique" stretch="normal" format="type1" metrics="/usr/share/fonts/type1/gsfo
nts/a010033l.afm" glyphs="/usr/share/fonts/type1/gsfonts/a010033l.pfb"/>
  <type name="AvantGarde-Demi" fullname="AvantGarde DemiBold" family="AvantGarde" foundry="URW" weight="600" style="normal" stretch="normal" format="type1" metrics="/usr/share/fonts/type1/gsfonts/a010015l
.afm" glyphs="/usr/share/fonts/type1/gsfonts/a010015l.pfb"/>
  ...
![FontBrowser](https://github.com/ImageMagick/ImageMagick/assets/153065254/1b9a534a-92b5-4cdb-b3f2-a7e890211cd5)

they were generated by

./configure \
  --without-perl      --with-gslib          --with-fontpath \
  --with-bzlib=yes    --with-autotrace=no   --with-djvu=yes \
  --with-dps=no       --with-fftw=no        --with-flif=no \
  --with-fpx=no       --with-fontconfig=yes --with-freetype=yes \
  --with-gslib=yes    --with-gvc=yes        --with-heic=yes \
  --with-jbig=yes     --with-jpeg=yes       --with-jxl=yes \
  --with-dmr=yes      --with-lcms=yes       --with-lqr=yes \
  --with-lzma=yes       --with-magick-plus-plus=yes \
  --with-openexr=yes  --with-openjp2=yes    --with-pango=yes \
  --with-perl=no      --with-png=yes        --with-raqm=yes \
  --with-raw=yes      --with-rsvg=no        --with-tiff=yes \
  --with-webp=yes     --with-wmf=no         --with-x=yes   \
  --with-xml=yes      --with-zip=yes        --with-zlib=yes \
  --with-zstd=yes \
  --with-dejavu-font-dir=/usr/share/fonts/truetype/dejavu/ \
  --with-gs-font-dir=/usr/share/fonts/type1/gsfonts  \
  --with-urw-base35-font-dir=/usr/share/fonts/type1/urw-base35

/usr/local/bin/convert -list font | grep glyphs | wc
    500    1000   33297

/usr/local/bin/identify -version
Version: ImageMagick 7.1.1-23 (Beta) Q16-HDRI x86_64 51291ec84:20231204 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI OpenMP(4.5) 
Delegates (built-in): fontconfig freetype gslib jbig jng jpeg lzma
pangocairo png ps tiff webp x zlib
Compiler: gcc (11.4)

the file is correct
/usr/local/lib/ImageMagick-7.1.1/config-Q16HDRI/configure.xml

identify -list format 
  give a complete list

  
/usr/local/bin/identify -list configure | grep DELEGATES
DELEGATES fontconfig freetype gslib jbig jng jpeg lzma pangocairo png
ps tiff video webp x zlib
DELEGATES fontconfig freetype gslib jbig jng jpeg lzma pangocairo png
ps tiff webp x zlib

/usr/local/bin/identify -list configure | grep LIBS
LIBS -ljbig -ltiff -lfreetype -ljpeg -lgs -L/usr/local/lib -lpng16
-lfontconfig -lfreetype -lwebpmux -lwebpdemux -lwebp -lwebp -lXext
-lXt -lSM -lICE -lX11 -llzma -lpangocairo-1.0 -lpango-1.0
-lgobject-2.0 -lglib-2.0 -lharfbuzz -lcairo -lz -lm -lpthread -lgomp

My previous version of ImageMagic worked fine on Ubuntu

help 

Kim

**Comment by fmw42** [comment]
Use the script at https://imagemagick.org/Usage/scripts/imagick_type_gen to regenerate the type.xml list from your fonts directory so that they can be used by name.

**Comment by KFH222** [comment]
On Wed, 2023-12-06 at 16:40 +0000, Fred Weinhaus wrote:
> Use the script at
> https://imagemagick.org/Usage/scripts/imagick_type_gen to regenerate
> the type.xml list from your fonts directory so that they can be used
> by name.
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.Message ID:
> ***@***.***
> b.com>


Thank you for your comment

However the problem as far as I see is that the configure
script for imagemagic relies on linux doing thr correct thing in
when installing something like libxml it puts in the correct
place in say /usr/lib/pkgconfig/  the file softare.pc

Centos I used since 1900's did so

but my desktop machine failed and I got a new machine and 
was refused to install centos (which is dead anyway)
and gave me a ubuntu machine which I have found to be
barely linux machine and in fact using the configure
under imagemagic as it cant find for example the libxml2.pc file
(when I have it installed ) it fails to load this

it does this for about 8 lib files which all I have
but there is no xx.pc file in the lib/pkgconfig directory

I am stuck that one of my most used programs 
imagemagic is now dead 

I am also a fan of your imagemagic scripts
http://www.fmwconcepts.com/imagemagick/


but the use of imagick_type_gen under ubuntu

fails to make correct type.xml files

e.g.

I get when I use 
./configure ...\
  --with-dejavu-font-dir=/usr/share/fonts/truetype/dejavu/ \
  --with-gs-font-dir=/usr/share/fonts/type1/gsfonts  \
  --with-urw-base35-font-dir=/usr/share/fonts/type1/urw-base35

I get e.g in type-ghostscript.xml

 <type name="AvantGarde-Book" fullname="AvantGarde Book"
family="AvantGarde" foundry="URW" weight="400" s
tyle="normal" stretch="normal" format="type1"
metrics="/usr/share/fonts/type1/gsfonts/a010013l.afm" glyphs
="/usr/share/fonts/type1/gsfonts/a010013l.pfb"/>

  etc

but using
imagick_type_gen

I get partly empty definitions and rathyer rubbish defintions

o/p example compare to above


  <type
     format="type1"
     name="URWGothicBook"
     fullname="URW Gothic Book"
     family="URW Gothic"
     glyphs="/usr/share/fonts/type1/gsfonts/a010013l.pfb"
     metrics="/usr/share/fonts/type1/gsfonts/a010013l.afm"
     />

i.e. default values for weight etc are missing 
and even the naming doesnt match

I am sad to say that one of my most used pieces of softwware is dead
under Ubuntu
  
I have libxml, bzip, espgs, libpng, all install in .usr/lib
but had to restall from source into /usr/local/
as the configure script in IM cant find them in the normal path
as no xxx.pc file is installed and installations in /snap
are invisible

This may be a problem in that the Install is for unix
and althought it indicates it is also for linux it is out of date
with snap and flatpack installations.

I have installed from scratch into /usr/local
and the now the xx.pc files are visible
with my PKG_CONFIG_LIBDIR   PKG_CONFIG_PATH  LD_LIBRARY_PATH

but still the install /usr/local IM 
display shows some sort of rubbish font list
when my installed source directories for
type.xml etc are 100% correct

the type.xml files are 100% correctv their locations are in the
correct places but Imagemagic 7.1.1.1 fails to recognise these files

Thanks but sad

(sorry also my original question raising my problem
the software for that mode deleted all the xml examples)


kim

**Reply by fmw42** [reply]
I am not an expert on Linux installations.  But

```
<type 
format="type1" 
name="URWGothicBook" 
fullname="URW Gothic Book" 
family="URW Gothic" 
glyphs="/usr/share/fonts/type1/gsfonts/a010013l.pfb" 
metrics="/usr/share/fonts/type1/gsfonts/a010013l.afm" 
/>

```

is the correct format when using that script.  It is what I get all the time. I have never seen weight in it.  The format for Ghostscript fonts may be different. I do not know much about that.

An IM developer will need to comment further.

**Comment by KFH222** [comment]
again thank you

i have decided by systems at fault not anything to do
with Imagemagic

as I install from source GraphicsMagick-1.3.42

Ths has type.mg instead of type.xml

and surprise surprise I have identical
annotation 'browser' menu as in ImageMacik

so my machine in  both sets of software is
showing the default menu and not reading
the installation files
  type .xml
or 
  type.mgk

so I have some software broken that they
both rely on to read the fonts

I have check everyother  type of software that uses 
font like libreoffice and they all work

so I should just give up buy redhat linux
and start all over again.

kim
